pkg/chaincode/signaturepolicy.go

```diff
@@ -391,3 +392,70 @@ func signaturePolicyEnvelopeFromString(policy string) (*cb.SignaturePolicyEnvelo
 
 	return p, nil
 }
+
+// SignaturePolicyEnvelopeToString parse a SignaturePolicyEnvelope to human readable expression
+// the returned expression is GATE(P[, P])
+//
+// where:
+//   - GATE is either "and" or "or" or "outof"
+//   - P is either a principal or another nested call to GATE
+//
+// A principal is defined as:
+//
+// # ORG.ROLE
+//
+// where:
+//   - ORG is a string (representing the MSP identifier)
+//   - ROLE takes the value of any of the RoleXXX constants representing
+//     the required role
+func SignaturePolicyEnvelopeToString(policy *cb.SignaturePolicyEnvelope) (string, error) {
+	ids := []string{}
+	for _, id := range policy.Identities {
+		var mspRole mb.MSPRole
+		err := proto.Unmarshal(id.Principal, &mspRole)
+		if err != nil {
+			return "", err
+		}
+		mspid := mspRole.MspIdentifier + "." + strings.ToLower(mb.MSPRole_MSPRoleType_name[int32(mspRole.Role)])
+		ids = append(ids, mspid)
+	}
+
+	var buf bytes.Buffer
+	policyParse(policy.Rule.Type, ids, &buf)
+	return buf.String(), nil
+}
+
+//recursive parse
+func policyParse(t any, ids []string, buf *bytes.Buffer) {
+	p1, ok := t.(*cb.SignaturePolicy_SignedBy)
+	if ok {
+		buf.WriteString("'")
+		buf.WriteString(ids[p1.SignedBy])
+		buf.WriteString("'")
+		return
+	}
+
+	p2, ok := t.(*cb.SignaturePolicy_NOutOf_)
+	if ok {
+		l := len(p2.NOutOf.Rules)
+		n := int32(l)
+
+		if p2.NOutOf.N == n {
+			buf.WriteString("AND(")
+		} else if p2.NOutOf.N == 1 {
+			buf.WriteString("OR(")
+		} else {
+			buf.WriteString("OutOf(")
+			buf.WriteString(fmt.Sprint(p2.NOutOf.N))
+			buf.WriteString(",")
+		}
+		for i, r := range p2.NOutOf.Rules {
+			policyParse(r.Type, ids, buf)
+			if i == len(p2.NOutOf.Rules)-1 {
+				buf.WriteString(")")
+			} else {
+				buf.WriteString(",")
+			}
+		}
+	}
+}
```

Comment:
One potential issue is that it looks like processing an empty slice of NOutOf.Rules, the closing `)` can be missed as the for-loop won't run. This might be a good case to unit test.

In general it's good practice to use the `Get...()` methods on protobuf messages instead of reading the fields directly as they avoid potential nil pointer issues.

Not a functional problem but from a readability and maintainability perspective, I am not very keen on:

1. Unnecessary use of `any`, losing any compile-time type checking to aid proper usage.
1. Variables scoped more broadly than they need to be, leading to names like `p1` and `p2`.
1. The early return where there is no "else" code in the function outside of any "if" statements, making the flow slightly less clear.

Here is a suggestion for an alternative style to consider:
```suggestion
func SignaturePolicyEnvelopeToString(policy *cb.SignaturePolicyEnvelope) (string, error) {
	ids := []string{}
	for _, id := range policy.GetIdentities() {
		var mspRole mb.MSPRole
		if err := proto.Unmarshal(id.GetPrincipal(), &mspRole); err != nil {
			return "", err
		}
		mspID := mspRole.GetMspIdentifier() + "." + strings.ToLower(mb.MSPRole_MSPRoleType_name[int32(mspRole.GetRole())])
		ids = append(ids, mspID)
	}

	var buf bytes.Buffer
	policyParse(policy.GetRule(), ids, &buf)
	return buf.String(), nil
}

// recursive parse
func policyParse(rule *cb.SignaturePolicy, ids []string, buf *bytes.Buffer) {
	switch p := rule.GetType().(type) {
	case *cb.SignaturePolicy_SignedBy:
		buf.WriteString("'")
		buf.WriteString(ids[p.SignedBy])
		buf.WriteString("'")

	case *cb.SignaturePolicy_NOutOf_:
		n := p.NOutOf.GetN()
		rules := p.NOutOf.GetRules()

		switch n {
		case int32(len(rules)):
			buf.WriteString("AND(")
		case 1:
			buf.WriteString("OR(")
		default:
			buf.WriteString("OutOf(")
			buf.WriteString(fmt.Sprint(n))
			buf.WriteString(",")
		}

		lastRuleIndex := len(rules) - 1
		for i, r := range rules {
			policyParse(r, ids, buf)
			if i < lastRuleIndex {
				buf.WriteString(",")
			}
		}
		buf.WriteString(")")
	}
}
```